cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs

```diff
@@ -353,70 +372,147 @@ readSigningKeyFile skFile =
+        AShelleyScriptWitness ssw -> go next rest (bwl, skwl, sswl ++ [ssw])
+
+data ByronOrShelleyWitness
+  = AByronWitness !(Api.SigningKey Api.ByronKey)
+  | AShelleyKeyWitness !Api.ShelleyWitnessSigningKey
+  | AShelleyScriptWitness !Api.MultiSigScript
+
+categoriseSomeWitness :: SomeWitness -> ByronOrShelleyWitness
+categoriseSomeWitness swsk =
   case swsk of
-    AByronSigningKey           sk -> Left sk
-    APaymentSigningKey         sk -> Right (Api.WitnessPaymentKey         sk)
-    APaymentExtendedSigningKey sk -> Right (Api.WitnessPaymentExtendedKey sk)
-    AStakeSigningKey           sk -> Right (Api.WitnessStakeKey           sk)
-    AStakeExtendedSigningKey   sk -> Right (Api.WitnessStakeExtendedKey   sk)
-    AStakePoolSigningKey       sk -> Right (Api.WitnessStakePoolKey       sk)
-    AGenesisSigningKey         sk -> Right (Api.WitnessGenesisKey sk)
-    AGenesisExtendedSigningKey sk -> Right (Api.WitnessGenesisExtendedKey sk)
-    AGenesisDelegateSigningKey sk -> Right (Api.WitnessGenesisDelegateKey sk)
+    AByronSigningKey           sk -> AByronWitness sk
+    APaymentSigningKey         sk -> AShelleyKeyWitness (Api.WitnessPaymentKey         sk)
+    APaymentExtendedSigningKey sk -> AShelleyKeyWitness (Api.WitnessPaymentExtendedKey sk)
+    AStakeSigningKey           sk -> AShelleyKeyWitness (Api.WitnessStakeKey           sk)
+    AStakeExtendedSigningKey   sk -> AShelleyKeyWitness (Api.WitnessStakeExtendedKey   sk)
+    AStakePoolSigningKey       sk -> AShelleyKeyWitness (Api.WitnessStakePoolKey       sk)
+    AGenesisSigningKey         sk -> AShelleyKeyWitness (Api.WitnessGenesisKey sk)
+    AGenesisExtendedSigningKey sk -> AShelleyKeyWitness (Api.WitnessGenesisExtendedKey sk)
+    AGenesisDelegateSigningKey sk -> AShelleyKeyWitness (Api.WitnessGenesisDelegateKey sk)
     AGenesisDelegateExtendedSigningKey sk
-                                  -> Right (Api.WitnessGenesisDelegateExtendedKey sk)
-    AGenesisUTxOSigningKey     sk -> Right (Api.WitnessGenesisUTxOKey     sk)
+                                  -> AShelleyKeyWitness (Api.WitnessGenesisDelegateExtendedKey sk)
+    AGenesisUTxOSigningKey     sk -> AShelleyKeyWitness (Api.WitnessGenesisUTxOKey     sk)
+    AShelleyMultiSigScript scr    -> AShelleyScriptWitness scr
 
 runTxGetTxId :: TxBodyFile -> ExceptT ShelleyTxCmdError IO ()
 runTxGetTxId (TxBodyFile txbodyFile) = do
   txbody <- firstExceptT ShelleyTxCmdReadTextViewFileError . newExceptT $
               Api.readFileTextEnvelope Api.AsShelleyTxBody txbodyFile
   liftIO $ BS.putStrLn $ Api.serialiseToRawBytesHex (Api.getTxId txbody)
 
-runTxWitness
+runTxCreateWitness
   :: TxBodyFile
-  -> SigningKeyFile
+  -> SigningKeyOrScriptFile
   -> Maybe NetworkId
   -> OutputFile
   -> ExceptT ShelleyTxCmdError IO ()
-runTxWitness (TxBodyFile txbodyFile) witSignKeyFile mbNw (OutputFile oFile) = do
+runTxCreateWitness (TxBodyFile txbodyFile) sKeyOrScript (Just nw) (OutputFile oFile) = do
```

Comment:
This function should be refactored similarly to how Duncan described in https://github.com/input-output-hk/cardano-node/pull/1788#discussion_r488868426.